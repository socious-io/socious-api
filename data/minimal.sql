CREATE TABLE public.users(
  id serial PRIMARY KEY,
  first_name varchar(70) DEFAULT NULL,
  username varchar(200) UNIQUE NOT NULL,
  name varchar(255) DEFAULT NULL,
  email varchar(200) UNIQUE NOT NULL,
  email_text varchar(255) DEFAULT NULL,
  mobile_countries_id int(11) DEFAULT NULL,
  phone varchar(255) DEFAULT NULL,
  wallet_address varchar(255) DEFAULT NULL,
  password text NOT NULL,
  remember_token text DEFAULT NULL,
  city text,
  city_ja mediumtext,
  description_search mediumtext,
  description_search_ja mediumtext,
  address_detail text,
  avatar text,
  cover_image text,
  expiry_date datetime DEFAULT NULL,
  active boo,
  activation_token varchar(255) DEFAULT NULL,
  mission text,
  bio mediumtext,
  profile_id varchar(200) DEFAULT NULL,
  view_as int(11) DEFAULT NULL,
  language varchar(255) DEFAULT NULL,
  my_conversation varchar(255) DEFAULT NULL,  
  email_verified_at datetime DEFAULT NULL,
  impact_score float NOT NULL DEFAULT '0',
  lock_otp datetime DEFAULT NULL,
  created_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  updated_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
  deleted_at datetime DEFAULT NULL
);
