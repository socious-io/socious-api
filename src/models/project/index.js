import * as write from './write.js'
import * as read from './read.js'
import * as questions from './questions.js'

export default {
  ...read,
  ...write,
  ...questions
}
