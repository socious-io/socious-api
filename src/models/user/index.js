import * as write from './write.js'
import * as read from './read.js'
import * as enums from './enums.js'
import * as additionals from './additionals.js'

export default {
  ...enums,
  ...read,
  ...write,
  ...additionals
}
