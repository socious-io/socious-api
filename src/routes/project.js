import Router from '@koa/router'
import Data, { validate } from '@socious/data'
import Project from '../models/project/index.js'
import Applicant from '../models/applicant/index.js'
import Notif from '../models/notification/index.js'
import Mission from '../models/mission/index.js'
import Offer from '../models/offer/index.js'
import Event from '../services/events/index.js'
import { paginate } from '../utils/middlewares/requests.js'
import { loginOptional, loginRequired } from '../utils/middlewares/authorization.js'
import { checkIdParams, projectPermission } from '../utils/middlewares/route.js'
import { PermissionError } from '../utils/errors.js'
import Analytics from '../services/analytics/index.js'
import { recommendProjectByProject } from '../services/recommender/index.js'
export const router = new Router()

router.get('/categories', async (ctx) => {
  ctx.body = { categories: await Project.jobCategories() }
})

router.get('/:id', loginOptional, checkIdParams, async (ctx) => {
  ctx.body = await Project.get(ctx.params.id, ctx.user.id)
})

router.get('/', loginOptional, paginate, async (ctx) => {
  ctx.body = await Project.all(ctx.paginate)
})

router.post('/', loginRequired, async (ctx) => {
  // only organizations allow to create projects
  if (ctx.identity.type !== Data.IdentityType.ORG) throw new PermissionError()

  await validate.ProjectSchema.validateAsync(ctx.request.body)
  ctx.body = await Project.insert(ctx.identity.id, ctx.request.body)
})

router.post('/update/:id', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await validate.ProjectSchema.validateAsync(ctx.request.body)
  ctx.body = await Project.update(ctx.params.id, ctx.request.body)
})

router.post('/update/:id/close', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  ctx.body = await Project.close(ctx.params.id)
})

router.get('/:id/questions', loginOptional, checkIdParams, async (ctx) => {
  ctx.body = {
    questions: await Project.getQuestions(ctx.params.id)
  }
})

router.post('/:id/questions', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await validate.QuestionSchema.validateAsync(ctx.request.body)
  ctx.body = await Project.addQuestion(ctx.params.id, ctx.request.body)
})

router.post('/update/:id/questions/:question_id', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await validate.QuestionSchema.validateAsync(ctx.request.body)
  ctx.body = await Project.updateQuestion(ctx.params.question_id, ctx.request.body)
})

router.post('/remove/:id/questions/:question_id', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await Project.removeQuestion(ctx.params.question_id)
  ctx.body = { message: 'success' }
})

router.post('/remove/:id', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await Project.remove(ctx.params.id)
  ctx.body = {
    message: 'success'
  }
})

router.get('/:id/applicants', loginRequired, paginate, checkIdParams, async (ctx) => {
  ctx.body = await Applicant.getByProjectId(ctx.params.id, ctx.paginate)
})

router.post('/:id/applicants', loginRequired, checkIdParams, async (ctx) => {
  await validate.ApplicantSchema.validateAsync(ctx.request.body)
  ctx.body = await Applicant.apply(ctx.params.id, ctx.user.id, ctx.request.body)
  const project = await Project.get(ctx.body.project_id)

  Event.push(Event.Types.NOTIFICATION, project.identity_id, {
    type: Notif.Types.APPLICATION,
    refId: ctx.body.id,
    parentId: project.id,
    identity: ctx.identity
  })

  Analytics.track({
    userId: ctx.params.user_id,
    event: 'applied_project',
    meta: ctx.body
  })
})

router.get('/:id/missions', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.paginate.filter.project_id = ctx.params.id
  ctx.paginate.filter.assigner_id = ctx.identity.id

  ctx.body = await Mission.getAll(ctx.paginate)
})

router.get('/:id/offers', loginRequired, checkIdParams, projectPermission, paginate, async (ctx) => {
  ctx.paginate.filter.project_id = ctx.params.id

  ctx.body = await Offer.getAll(ctx.identity.id, ctx.paginate)
})

router.post('/:id/offer/:user_id', loginRequired, checkIdParams, projectPermission, async (ctx) => {
  await validate.OfferSchema.validateAsync(ctx.request.body)

  ctx.body = await Offer.send(ctx.params.id, {
    ...ctx.request.body,
    recipient_id: ctx.params.user_id,
    offerer_id: ctx.identity.id
  })

  Event.push(Event.Types.NOTIFICATION, ctx.params.user_id, {
    type: Notif.Types.OFFER,
    refId: ctx.body.id,
    parentId: ctx.params.id,
    identity: ctx.identity
  })

  Analytics.track({
    userId: ctx.params.user_id,
    event: 'offered_project',
    meta: ctx.project
  })
})

router.get('/:id/feedbacks', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.body = await Mission.feedbacks(ctx.params.id, ctx.paginate)
})

router.get('/:id/similars', loginOptional, checkIdParams, paginate, async (ctx) => {
  ctx.body = await recommendProjectByProject(ctx.params.id)
})
