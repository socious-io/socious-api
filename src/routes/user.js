import Router from '@koa/router'
import User from '../models/user/index.js'
import Applicant from '../models/applicant/index.js'
import Auth from '../services/auth/index.js'
import Mission from '../models/mission/index.js'
import Offer from '../models/offer/index.js'
import Skill from '../models/skill/index.js'
import ImpactPoints from '../services/impact_points/index.js'
import Payment from '../services/payments/index.js'
import { paginate } from '../utils/middlewares/requests.js'
import { loginOptional, loginRequired } from '../utils/middlewares/authorization.js'
import Data, { validate } from '@socious/data'
import { checkIdParams } from '../utils/middlewares/route.js'
import { putContact } from '../services/sendgrid/index.js'
import { BadRequestError } from '../utils/errors.js'
import { recommendUserByUser, recommendProjectByUser, recommendOrgByUser } from '../services/recommender/index.js'

export const router = new Router()

router.get('/:id/profile', loginOptional, checkIdParams, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileLimited(ctx.params.id)
    return
  }
  ctx.body = await User.getProfile(ctx.params.id, ctx.identity.id)
})

router.post('/:id/report', loginRequired, async (ctx) => {
  await validate.ReportSchema.validateAsync(ctx.request.body)
  await User.report({
    ...ctx.request.body,
    user_id: ctx.params.id,
    identity_id: ctx.identity.id
  })

  ctx.body = {
    message: 'success'
  }
})

router.get('/by-username/:username/profile', loginOptional, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileByUsernameLimited(ctx.params.username)
    return
  }
  ctx.body = await User.getProfileByUsername(ctx.params.username, ctx.identity.id)
})

router.get('/profile', loginRequired, async (ctx) => {
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/wallet', loginRequired, async (ctx) => {
  const wallet = ctx.request.body.wallet_address
  if (!wallet) throw new BadRequestError('wallet address is required')
  await User.updateWalletAddress(ctx.user.id, wallet)
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/profile', loginRequired, async (ctx) => {
  await validate.UpdateProfileSchema.validateAsync(ctx.request.body)

  const skills = await Skill.getAllByNames(ctx.request.body.skills)
  ctx.request.body.skills = skills.map((s) => s.name)

  ctx.body = await User.updateProfile(ctx.user.id, ctx.request.body)

  putContact({
    first_name: ctx.body.first_name,
    last_name: ctx.body.last_name,
    email: ctx.body.email
  })
})

router.post('/change-password', loginRequired, async (ctx) => {
  await Auth.changePassword(ctx.user, ctx.request.body)
  ctx.body = {
    message: 'success'
  }
})

router.post('/change-password-direct', loginRequired, async (ctx) => {
  await Auth.directChangePassword(ctx.user, ctx.request.body)

  ctx.body = {
    message: 'success'
  }
})

router.post('/delete', loginRequired, async (ctx) => {
  await User.remove(ctx.user, ctx.request.body.reason)
  ctx.body = {
    message: 'success'
  }
})

router.get('/applicants', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.body = await Applicant.getByUserId(ctx.user.id, ctx.paginate)
})

router.get('/:id/missions', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.params.id
  ctx.body = await Mission.getAll(ctx.paginate)
})

router.get('/missions', loginRequired, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.identity.id

  const missions = await Mission.getAll(ctx.paginate)

  ctx.body = missions.map((m) => {
    return {
      ...m,
      ...Payment.amounts({
        amount: m.offer.assignment_total,
        service: Data.PaymentService.STRIPE
          ? m.offer.payment_mode === Data.PaymentMode.FIAT
          : Data.PaymentService.CRYPTO,
        verified: m.assigner.meta.verified_impact
      })
    }
  })
})

router.get('/offers', loginRequired, paginate, async (ctx) => {
  ctx.body = await Offer.getAll(ctx.identity.id, ctx.paginate)
})

router.post('/languages', loginRequired, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addLanguage(ctx.user, ctx.request.body)
})

router.post('/languages/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editLanguage(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/languages/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  await User.removeLanguage(ctx.params.id, ctx.user)
  ctx.body = { message: 'success' }
})

router.post('/experiences', loginRequired, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addExperience(ctx.user, ctx.request.body)
})

router.post('/experiences/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editExperience(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/experiences/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await User.removeExperience(ctx.params.id, ctx.user)
})

router.get('/badges', loginRequired, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.identity.id) }
})

router.get('/:id/badges', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.params.id) }
})

router.get('/impact-points', loginRequired, paginate, async (ctx) => {
  ctx.body = await ImpactPoints.history(ctx.identity.id, ctx.paginate)
})

router.post('/open-to-work', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_work: await User.openToWork(ctx.user.id)
  }
})

router.post('/open-to-volunteer', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_volunteer: await User.openToVolunteer(ctx.user.id)
  }
})

router.get('/:username/recommend/jobs', loginOptional, async (ctx) => {
  ctx.body = await recommendProjectByUser(ctx.params.username)
})

router.get('/:username/recommend/users', loginOptional, async (ctx) => {
  ctx.body = await recommendUserByUser(ctx.params.username)
})

router.get('/:username/recommend/orgs', loginOptional, async (ctx) => {
  ctx.body = await recommendOrgByUser(ctx.params.username)
})
