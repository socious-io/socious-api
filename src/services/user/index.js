import * as write from './write.js';
import * as read from './read.js';
import * as auth from './auth.js';

export default {
  ...auth,
  ...read,
  ...write,
};
