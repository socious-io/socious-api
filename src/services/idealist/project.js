import axios from 'axios';
import https from 'https';
import http from 'http';

import sql from 'sql-template-tag';
import {app} from './app.js';

//import sql from 'sql-template-tag';
//import {app} from '../../../src/index.js';

import {organizationFromProject} from './projectOrganization.js';
import project from '../../models/project/index.js';

/**
 * Get single project from Idealist
 *
 * @param {string} project_types
 * @param {string} id
 * @returns object || void
 */
export const getProject = async function (project_types, id) {
  try {
    let proj = await axios.get(
      `https://www.idealist.org/api/v1/listings/${project_types}/${id}`,
      {
        auth: {
          username: '743e1f3940484d7680130c748ed22758',
          password: '',
        },

        timeout: 0,
        httpAgent: new http.Agent({keepAlive: true}),
        httpsAgent: new https.Agent({keepAlive: true}),
      },
    );

    if (proj.status === 200) {
      return proj.data;
    } else {
      console.log('\x1b[31m%s\x1b[0m', proj.status + ', ' + proj.statusText);
    }
  } catch (err) {
    console.log('\x1b[31m%s\x1b[0m', err.message);
  }
};

export const lastIdealistProject = async function (project_type) {
  try {
    project_type = project_type.slice(0, -1);

    const row = await app.db.get(
      sql`SELECT updated_at FROM projects WHERE other_party_title = ${project_type} ORDER BY updated_at DESC LIMIT 1`,
    );

    return new Date(row.updated_at).toISOString();
  } catch (err) {
    if (err.status !== 400)
      console.log('\x1b[31m%s\x1b[0m', err.status + ' ' + err.message);
    return null;
  }
};

export const processProject = async function (p, type) {
  try {
    let org_id;

    //create organization or get ID of existing one
    org_id = await organizationFromProject(p);

    //if project is volop, can have a group (create or find and get ID)
    // if (project.group) {
    //   group_id = await groupFromProject(project.group);
    // }

    if (org_id) return await saveProject(p, type, org_id); //return true or false
  } catch (err) {
    console.log('\x1b[31m%s\x1b[0m', err.message);
  }
};

async function saveProject(pro, type, org_id) {
  try {
    const payment_type = await paymentType(pro);
    const remote_preference = await remotePreference(pro);

    const payment_scheme =
      pro.salaryPeriod && pro.salaryPeriod == 'HOUR' ? 'HOURLY' : 'FIXED';
    const experience_level = await experienceLevel(pro);

    const body = {
      title: pro.name ? pro.name : 'Untitled',
      description: pro.description ? pro.description : 'No information',
      ...(pro.address && pro.address.country && {country: pro.address.country}),
      remote_preference: remote_preference,
      payment_type: payment_type,
      payment_scheme: payment_scheme,
      ...(pro.salaryCurrency && {payment_currency: pro.salaryCurrency}),
      ...(pro.salaryMinimum && {
        payment_range_lower: pro.salaryMinimum.toString(),
      }),
      ...(pro.salaryMaximum && {
        payment_range_higher: pro.salaryMaximum.toString(),
      }),
      ...(experience_level && {experience_level: experience_level}),
      status: 'ACTIVE',
      ...(pro.expires && {expires_at: pro.expires}), //check if works for date
      other_party_id: pro.id,
      other_party_title: type,
      ...(pro.url && pro.url.en && {other_party_url: pro.url.en}),
      updated_at: pro.updated,
      // updated_at is important to find the last downloaded project from Idealist
      // they are sorted by updated_at on their listings and it is used into the 'since' param
      // when getting the listings
    };

    // insert or update the project in DB
    const projFromDb = await getProjectFromDb(pro);

    if (projFromDb) {
      //update project
      await project.update(projFromDb.id, body);
    } else {
      //create new project
      await project.insert(org_id, body);
    }

    return true;
  } catch (err) {
    console.log('\x1b[31m%s\x1b[0m', err.message, err);
    return false;
  }
}

async function getProjectFromDb(p) {
  try {
    const pr = await app.db.get(
      sql`SELECT id FROM projects WHERE other_party_id = ${p.id}`,
    );

    return pr.id;
  } catch (err) {
    if (err.status !== 400) console.log(err.message, err.status);
    return false;
  }
}

// PARSING FUNCTIONS FOR PROJECT

/**
 * get payment type
 *
 * @param object
 * @return string
 */
async function paymentType(p) {
  if (
    (p.salaryMinimum && p.salaryMinimum > 0) ||
    (p.salaryMaximum && p.salaryMaximum > 0) ||
    (p.paid && p.paid == true)
  ) {
    return 'PAID';
  } else {
    return 'VOLUNTEER';
  }
}

/**
 * get project remote preference
 *
 * @param object
 * @return string
 */
async function remotePreference(p) {
  if (p.remoteOk && p.remoteOk === true) {
    if (p.remoteTemporary && p.remoteTemporary === true) {
      return 'HYBRID';
    }
    return 'REMOOTE'; //change to enum and here to REMOTE!!!
  } else {
    return 'ONSITE';
  }
}

/**
 * project experience level
 *
 * @param object
 * @return integer
 */
async function experienceLevel(p) {
  if (p.professionalLevel) {
    switch (p.professionalLevel) {
      case 'NONE':
        return 1;
      case 'ENTRY_LEVEL':
        return 2;
      case 'MANAGERIAL':
      case 'INTERMEDIATE':
        return 3;
      case 'PROFESSIONAL':
      case 'DIRECTOR':
      case 'EXECUTIVE':
      case 'EXPERT':
        return 4;
      default:
        return 0;
    }
  }
  return 0;
}
