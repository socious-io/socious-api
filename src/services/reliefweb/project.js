import sql from 'sql-template-tag'
import { app } from '../../index.js'
import project from '../../models/project/index.js'
import { createOrgFromProject } from './organization.js'
import { iso3ToIso2 } from './helpers.js'

const other_party_title = 'reliefweb job'

/**
 *
 * @example
 */
export async function getLastReliefWebProjectId() {
  try {
    const row = await app.db.get(
      sql`SELECT other_party_id FROM projects WHERE other_party_title = ${other_party_title} 
      ORDER BY other_party_id DESC LIMIT 1`
    )

    return row.other_party_id || 0
  } catch (err) {
    if (err.status !== 400 && err.message !== 'Not matched') {
      console.log('\x1b[31m%s\x1b[0m', err.status + ' ' + err.message)
    }

    return 0
  }
}

/**
 *
 * @param p
 * @example
 */
export async function processProject(p) {
  const orgId = await createOrgFromProject(p)

  if (!orgId) return false

  // save or update the project
  return await saveProject(p, other_party_title, orgId)
}

/**
 *
 * @param p - Project.
 * @param org_id
 * @param type
 * @param orgId
 * @returns Boolean.
 * @example
 */
async function saveProject(p, type, orgId) {
  try {
    const body = await parseProject(p, type)

    // insert or update the project in DB
    const projFromDb = await getProjectFromDb(p)

    if (projFromDb) {
      await project.update(projFromDb.id, body)
    } else {
      await project.insert(orgId, body)
    }

    return true
  } catch (err) {
    console.log('\x1b[31m%s\x1b[0m', err.message, err)
    return false
  }
}

/**
 *
 * @param p
 * @param type
 * @example
 */
async function parseProject(p, type) {
  if (!p.title) return false

  const experienceLevel = await getExperienceLevel(p)
  let countryIso2 = null
  if (p.country && p.country[0].iso3) {
    countryIso2 = await iso3ToIso2(p.country[0].iso3)
  }

  return {
    title: p.title,
    description: p['body-html'] ? p['body-html'] : p.body ? p.body : 'No information aviable',
    country: countryIso2,
    city: p.city[0].name,
    experience_level: experienceLevel,
    status: 'ACTIVE',
    expires_at: p.date.closing,
    other_party_id: p.id,
    other_party_title: type,
    other_party_url: p.url,
    updated_at: p.date.changed
  }
}

/**
 *
 * @param p
 * @example
 */
async function getExperienceLevel(p) {
  if (p.experience && p.experience[0].name) {
    const exp = p.experience[0].name
    let secondNum, firstNum
    if (Array.from(exp)[2]) {
      firstNum = parseInt(Array.from(exp)[2])
    }
    if (Array.from(exp)[0]) {
      secondNum = parseInt(Array.from(exp)[0])
    }
    if (!isNaN(firstNum) && !isNaN(secondNum)) {
      const res = Math.round((firstNum + secondNum) / 2)

      return res < 4 ? res : 4
    }
    return 0
  }
}

/**
 *
 * @param p
 * @example
 */
async function getProjectFromDb(p) {
  try {
    const pr = await app.db.get(sql`SELECT id FROM projects WHERE other_party_id = ${p.id}`)

    return pr.id
  } catch (err) {
    if (err.status !== 400 && err.message !== 'Not matched') {
      console.log(err.message, err.status)
    }
    return false
  }
}
