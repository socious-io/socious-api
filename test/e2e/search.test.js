import supertest from 'supertest'
import { app } from '../../src/index.js'
import data from '../data/index.js'
import { registerAndVerify } from './globals/user.js'
import { create as createOrg } from './globals/org.js'
import { create as createProject } from './globals/project.js'

let server, request

beforeAll(async () => {
  server = app.listen()
  request = supertest(server)

  await registerAndVerify(request, data)
  await createOrg(request, data)
  await createProject(request, data)
})

test('search projects', async () => {
  const response = await request.post('/search').set('Authorization', data.users[0].access_token).send({
    q: 'test',
    type: 'projects'
  })
  expect(response.status).toBe(200)
  expect(response.body).toMatchSnapshot({
    items: [
      {
        created_at: expect.any(String),
        updated_at: expect.any(String),
        id: expect.any(String),
        identity_id: expect.any(String),
        identity_meta: expect.any(Object),
        job_category_id: expect.any(String),
        job_category: expect.any(Object)
      },
      {
        created_at: expect.any(String),
        updated_at: expect.any(String),
        id: expect.any(String),
        identity_id: expect.any(String),
        identity_meta: expect.any(Object),
        job_category_id: expect.any(String),
        job_category: expect.any(Object)
      }
    ]
  })
})

test('projects with filter', async () => {
  const response = await request
    .post('/search')
    .set('Authorization', data.users[0].access_token)
    .send({
      type: 'projects',
      filter: {
        payment_currency: 'USD'
      }
    })
  expect(response.status).toBe(200)
  expect(response.body).toMatchSnapshot({
    items: [
      {
        created_at: expect.any(String),
        updated_at: expect.any(String),
        id: expect.any(String),
        identity_id: expect.any(String),
        identity_meta: expect.any(Object),
        job_category_id: expect.any(String),
        job_category: expect.any(Object)
      }
    ]
  })
})

test('search users', async () => {
  const response = await request.post('/search').set('Authorization', data.users[0].access_token).send({
    q: ' ',
    type: 'related_users'
  })
  expect(response.status).toBe(200)
})

test('search related users', async () => {
  const response = await request.post('/search').set('Authorization', data.users[0].access_token).send({
    q: ' ',
    type: 'users'
  })
  expect(response.status).toBe(200)
})

const cleanup = async () => {
  await app.db.query('DELETE FROM users')
  await app.db.query('DELETE FROM organizations')
  await app.db.query('DELETE FROM projects')
  await app.db.pool.end()
}

afterAll((done) => {
  cleanup(done).then(() => {
    server.close(done)
  })
})
