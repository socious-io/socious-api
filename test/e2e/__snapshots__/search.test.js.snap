// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
{
  "error": ""email" must be a valid email",
}
`;

exports[` 2`] = `
{
  "address": "address",
  "bio": "test",
  "city": "Tokyo",
  "country": "JP",
  "cover_image": null,
  "created_at": Any<String>,
  "created_by": Any<String>,
  "culture": "test",
  "description": "test",
  "email": "test@socious.io",
  "followers": 0,
  "followings": 0,
  "geoname_id": null,
  "hiring": false,
  "id": Any<String>,
  "image": null,
  "impact_points": 0,
  "industry": null,
  "mission": "test",
  "mobile_country_code": null,
  "name": "test ORG",
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "phone": "9379911",
  "search_tsv": Any<String>,
  "shortname": Any<String>,
  "size": null,
  "social_causes": [
    "SOCIAL",
    "HEALTH",
  ],
  "status": "ACTIVE",
  "type": "COOP",
  "updated_at": Any<String>,
  "verified_impact": false,
  "wallet_address": null,
  "website": "http://test.com",
}
`;

exports[` 3`] = `
{
  "address": "address",
  "bio": "test",
  "city": "Tehran",
  "country": "IR",
  "cover_image": null,
  "created_at": Any<String>,
  "created_by": Any<String>,
  "culture": "test",
  "description": "test",
  "email": "test@socious.io",
  "followers": 0,
  "followings": 0,
  "geoname_id": null,
  "hiring": false,
  "id": Any<String>,
  "image": null,
  "impact_points": 0,
  "industry": null,
  "mission": "test",
  "mobile_country_code": null,
  "name": "test ORG",
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "phone": "9899993",
  "search_tsv": Any<String>,
  "shortname": Any<String>,
  "size": null,
  "social_causes": [
    "SOCIAL",
    "MENTAL",
  ],
  "status": "ACTIVE",
  "type": "SOCIAL",
  "updated_at": Any<String>,
  "verified_impact": false,
  "wallet_address": null,
  "website": "http://test.com",
}
`;

exports[` 4`] = `
{
  "error": ""phone" must be a string",
}
`;

exports[` 5`] = `
{
  "causes_tags": [
    "POVERTY",
  ],
  "city": "Tokyo",
  "commitment_hours_higher": "13",
  "commitment_hours_lower": "12",
  "country": "JP",
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "Data testing",
  "experience_level": 1,
  "expires_at": null,
  "geoname_id": null,
  "id": Any<String>,
  "identity_id": Any<String>,
  "job_category_id": Any<String>,
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "payment_currency": "JPY",
  "payment_range_higher": "13",
  "payment_range_lower": "12",
  "payment_scheme": "FIXED",
  "payment_type": "VOLUNTEER",
  "project_length": "1_3_MONTHS",
  "project_type": "FULL_TIME",
  "remote_preference": "HYBRID",
  "search_tsv": Any<String>,
  "skills": [
    "ANDROID",
    "ANGULARJS",
  ],
  "status": "ACTIVE",
  "title": "Testing 1",
  "updated_at": Any<String>,
  "weekly_hours_higher": "13",
  "weekly_hours_lower": "12",
}
`;

exports[` 6`] = `
{
  "causes_tags": [
    "POVERTY",
  ],
  "city": "Tehran",
  "commitment_hours_higher": "13",
  "commitment_hours_lower": "12",
  "country": "IR",
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "Data testing",
  "experience_level": 1,
  "expires_at": null,
  "geoname_id": null,
  "id": Any<String>,
  "identity_id": Any<String>,
  "job_category_id": Any<String>,
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "payment_currency": "USD",
  "payment_range_higher": "13",
  "payment_range_lower": "12",
  "payment_scheme": "FIXED",
  "payment_type": "VOLUNTEER",
  "project_length": "LESS_THAN_A_MONTH",
  "project_type": "ONE_OFF",
  "remote_preference": "HYBRID",
  "search_tsv": Any<String>,
  "skills": [
    "ANDROID",
    "ANGULARJS",
  ],
  "status": "ACTIVE",
  "title": "Testing 2",
  "updated_at": Any<String>,
  "weekly_hours_higher": "13",
  "weekly_hours_lower": "12",
}
`;

exports[`projects with filter 1`] = `
{
  "items": [
    {
      "applicants": 0,
      "causes_tags": [
        "POVERTY",
      ],
      "city": "Tehran",
      "commitment_hours_higher": "13",
      "commitment_hours_lower": "12",
      "country": "IR",
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "Data testing",
      "experience_level": 1,
      "expires_at": null,
      "geoname_id": null,
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_type": "organizations",
      "job_category": Any<Object>,
      "job_category_id": Any<String>,
      "missions": 0,
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "payment_currency": "USD",
      "payment_range_higher": "13",
      "payment_range_lower": "12",
      "payment_scheme": "FIXED",
      "payment_type": "VOLUNTEER",
      "project_length": "LESS_THAN_A_MONTH",
      "project_type": "ONE_OFF",
      "remote_preference": "HYBRID",
      "search_tsv": "'2':2 'data':3 'test':1,4",
      "skills": [
        "ANDROID",
        "ANGULARJS",
      ],
      "status": "ACTIVE",
      "title": "Testing 2",
      "updated_at": Any<String>,
      "weekly_hours_higher": "13",
      "weekly_hours_lower": "12",
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 1,
}
`;

exports[`search projects 1`] = `
{
  "items": [
    {
      "applicants": 0,
      "causes_tags": [
        "POVERTY",
      ],
      "city": "Tehran",
      "commitment_hours_higher": "13",
      "commitment_hours_lower": "12",
      "country": "IR",
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "Data testing",
      "experience_level": 1,
      "expires_at": null,
      "geoname_id": null,
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_type": "organizations",
      "job_category": Any<Object>,
      "job_category_id": Any<String>,
      "missions": 0,
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "payment_currency": "USD",
      "payment_range_higher": "13",
      "payment_range_lower": "12",
      "payment_scheme": "FIXED",
      "payment_type": "VOLUNTEER",
      "project_length": "LESS_THAN_A_MONTH",
      "project_type": "ONE_OFF",
      "remote_preference": "HYBRID",
      "search_tsv": "'2':2 'data':3 'test':1,4",
      "skills": [
        "ANDROID",
        "ANGULARJS",
      ],
      "status": "ACTIVE",
      "title": "Testing 2",
      "updated_at": Any<String>,
      "weekly_hours_higher": "13",
      "weekly_hours_lower": "12",
    },
    {
      "applicants": 0,
      "causes_tags": [
        "POVERTY",
      ],
      "city": "Tokyo",
      "commitment_hours_higher": "13",
      "commitment_hours_lower": "12",
      "country": "JP",
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "Data testing",
      "experience_level": 1,
      "expires_at": null,
      "geoname_id": null,
      "id": Any<String>,
      "identity_id": Any<String>,
      "identity_meta": Any<Object>,
      "identity_type": "organizations",
      "job_category": Any<Object>,
      "job_category_id": Any<String>,
      "missions": 0,
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "payment_currency": "JPY",
      "payment_range_higher": "13",
      "payment_range_lower": "12",
      "payment_scheme": "FIXED",
      "payment_type": "VOLUNTEER",
      "project_length": "1_3_MONTHS",
      "project_type": "FULL_TIME",
      "remote_preference": "HYBRID",
      "search_tsv": "'1':2 'data':3 'test':1,4",
      "skills": [
        "ANDROID",
        "ANGULARJS",
      ],
      "status": "ACTIVE",
      "title": "Testing 1",
      "updated_at": Any<String>,
      "weekly_hours_higher": "13",
      "weekly_hours_lower": "12",
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 2,
}
`;
